[decisions] keyname (#4715)
ElektraInitiative/libelektra: doc/decisions/3_decided/keyname.md

The escaped name is a more complicated representation as a `\0`-terminated string that maps 1:1 onto unescaped names.
More details can be found in [the relevant docs](/doc/KEYNAMES.md).

There is a conflict between these two forms in terms of API convenience and efficient execution.
Generally, the escaped form is more convenient to use, since it is entirely human-readable and just a "normal" string.
Implementing many tasks (like order comparisons) is, however, much simpler when using the unescaped name.
Additionally, using the unescaped name often results in more performant code as well.

A particularly common example that highlights the difficulties in handling escaped names is splitting the name into parts.
In the escaped name, this task requires correct handling of escape sequences, whereas in the unescaped name parts are always delimited by a `\0` byte.

Of course a naive way to solve this dilemma is to say "why not both" and store both versions.
However, this can result in a lot of memory use, so it is not immediately the best option.

The question now is, which representations should be used by `libelektra-core` and how.

## Constraints

> - Because `KeySet` is ordered by name and stores `Key`, the order comparison between the name of two `Key`s must be fast.

Author: Do we have any data, that shows anything slower than a `memcmp` is too slow?

If not, I'd rather leave the constraint as "must be fast", but maybe we add an assumption that it must be comparable to a single `memcmp`.

- We need a single pointer to a single buffer that contains the entire name of a `Key`.
  While there are other options, some of which could even save memory (e.g., split into parts and deduplicate), much of the `KeySet` internals rely on the fact that the name is a single buffer.
  Changing this would require major redesigns.

## Assumptions

> - In most cases the escaped name is used for developer convenience and not because of actual requirements.

Author: Hm, yeah I'll rephrase it slightly. It certainly is for convenience, but probably not for developers.

- The most common requirement for using the escaped name is UI: reading names from or displaying them in a user interface (e.g., `kdb` CLI)

## Considered Alternatives

### Only escaped name

Because the escaped name is a simple `\0`-terminated string, it can be represented as a single `char *`.

Storing the name as a single `char *` would be the most space efficient.
But resizing would require counting the length every time.
Therefore, for storage the better solution may be a `char *` and a `size_t`.

However, in the API the name could always be a single `char *`, making for a very easy to use API.

> The biggest problem with this approach is that comparing two escaped names is not trivial.
> The comparison needs to account for namespaces, parts and escaping.

Author: ```suggestion
The biggest problem with this approach is that comparing two escaped names is not trivial.
The comparison needs to account for namespaces, parts and escaping.
Previous benchmarks showed that it is very hard or even impossible to make the comparison of escaped names fast enough for our use cases.
```

IMO "not trivial" is the root problem. The previous benchmarks also can only show that the one implementation we had was too slow. Benchmarks cannot show that it is impossible to find a different faster algorithm.

### Only unescaped name

The unescaped name contains `\0` bytes.
It therefore must be represented as a pointer and a size.

This can make for less convenient API, but there are mitigation strategies using additional types.
Using unescaped names in code can be inconvenient, especially regarding the namespace.
Without a namespace a name could be written as e.g., `"foo\0bar"`.
But with a namespace it would be something like `"\1\0\foo\0bar"` and developers would need to remember what namespace `\1` is.
Using the `KEY_NS_*` constants like this is not easily possible.

Both order and hierarchy comparisons are very simple in this case and can be implemented with a single `memcmp` and a tiny amount of extra logic (e.g., to handle cascading names).
Iterating over the individual parts is also trivial, since all parts are separated by `\0` bytes.

### Only unescaped name, with separate namespace

In the above solution, the entire unescaped name (including the namespace) would always be considered one unit.
As such, there would only be a single pointer and a size in an API that needs a name.
This can be inconvenient, because it makes using the `KEY_NS_*` constants more difficult.

> This solution enhances the above, by considering the namespace a separate thing.

Author: I mean that's what the following sentences are meant to explain....

Not necessarily all the way separate as in #4732, but separate enough that the passing namespace and rest of name as separate arguments is allowed.

I'll try to explain it a bit better, and also incorporate some of #4732.

This means the escaped name will be stored for many keys and therefore the benefit of the lazy allocation is negated.

Without the lazy allocation benefit, the only difference to "Both escaped and unescaped name" is that we have fewer pointers and sizes in structs.
This saves some amount of memory and allocations, but makes internal code more difficult to write and understand.

## Decision

Go with "Only unescaped name, with separate namespace" from above:

- Store only unescaped name inside `struct _Key`
- API of `libelektra-core` will use unescaped name exclusively
- Convenience functions using escaped names, will be provided via other libraries
- Where appropriate the API will take the namespace as a separate argument to allow using `KEY_NS_*` constants.
- Whether namespace is stored separately in `struct _Key` will be decided at a later point, when the scope of all API changes and changes to `struct _Key` is clear.

## Rationale

- Largest memory savings among the proposed options
- Option to use separate namespace argument leads to more convenient API (`KEY_NS_*` constants).
- Simple internal code

> - Escaped name requirements can easily be solved by additional library (e.g., `libelektra-ease`), because not every caller will need those.

Author: > Probably better operations or even a dedicated library for escaping (if it is big enough).

IMO `libelektra-ease` makes more sense than `libelektra-extra`. Since escaped names should mostly be used for UI things, I think this would fit the descriptions in the `library_split.md` decision. I see the escaped names closer to the SHA256 Hash of a `KeySet` than something like a `ksCut` operation.

A completely separate library could make sense too. But it would be a pretty small library. The API would basically be just 4 functions:

```
isValidEscapedName()
escapeName()
cannonicalizeEscapedName()
unescapeName()
```

All together the code for those function would be only about 500-600 lines, based on the current code in `keyname.c`. So I certainly wouldn't say it is "big enough", but you could argue it is an important enough standalone use case to have it's own library. If we expect many people to need escaped names, but otherwise can just use `libelektra-core` then a standalone library certainly makes sense.

> Bindings should be required to also implement functions of this library (similar to `kdb`).

That would also be a reason to make it a standalone library. Otherwise it might be awkward to require that bindings implement just a part of a libraries API, but the rest is optional.

# Namespace and Name of Keys

## Problem

A `Key` in Elektra is identified by its name, which consists of a namespace and a number of name parts.
There are two common representations for the name: escaped and unescaped.

The unescaped form is essentially a single namespace byte plus and arbitrary-length sequence of arbitrary bytes, in which a `\0` byte separates parts.

> The escaped name is a more complicated representation as a `\0`-terminated string that maps 1:1 onto unescaped names.

Reviewer: Sentence hard to read:

```suggestion
The escaped name is a `\0`-terminated string that maps 1:1 onto unescaped names.
```

More details can be found in [the relevant docs](/doc/KEYNAMES.md).

There is a conflict between these two forms in terms of API convenience and efficient execution.
Generally, the escaped form is more convenient to use, since it is entirely human-readable and just a "normal" string.
Implementing many tasks (like order comparisons) is, however, much simpler when using the unescaped name.
Additionally, using the unescaped name often results in more performant code as well.

A particularly common example that highlights the difficulties in handling escaped names is splitting the name into parts.
In the escaped name, this task requires correct handling of escape sequences, whereas in the unescaped name parts are always delimited by a `\0` byte.

> Of course a naive way to solve this dilemma is to say "why not both" and store both versions.

Reviewer: ```suggestion
Before this decision, we stored both versions in every `Key`.
```

> However, this can result in a lot of memory use, so it is not immediately the best option.

Reviewer: ```suggestion
However, this resulted in too much memory use, so we need to find another solution.
```

The question now is, which representations should be used by `libelektra-core` and how.

## Constraints

- Because `KeySet` is ordered by name and stores `Key`, the order comparison between the name of two `Key`s must be "fast enough".
  (see assumption below)
- We need a single pointer to a single buffer that contains the entire name of a `Key`.
  While there are other options, some of which could even save memory (e.g., split into parts and deduplicate), much of the `KeySet` internals rely on the fact that the name is a single buffer.
  Changing this would require major redesigns.

## Assumptions

- In most cases the escaped name is used for convenience and not because of actual requirements.
- The most common case for using the escaped name is UI: reading names from or displaying them in a user interface (e.g., `kdb` CLI)
- In the constraint about order comparisons above, we assume that "fast enough" means "comparable to a single `memcmp`".

>   This is an assumption, because we do not have data showing this is a necessary requirement.

Reviewer: ```suggestion
  In profiling non-`memcmp`-based implementations, up to now, it was obvious that most of the CPU time is spent on the comparisons in several benchmarks (like adding keys to a keyset).
```

Author: I don't like your suggestion. What is "non-`memcmp`-based"? Where are those benchmarks? If we can't link them here it's still "we don't have the data".

How about:

```suggestion
  Profiling for previous implementations, not based on a single `memcmp` of unescaped names, showed the comparison as a bottleneck, while the current single-`memcmp` implementation does not show the bottleneck.
```

>   It may be possible to find a slower solution that is still "fast enough".

Reviewer: ```suggestion
  It may be nevertheless possible to find a slower solution that is still "fast enough".
```

Author: ```suggestion
  That said, it may be possible to find a solution slower than the current one that is still fast enough to avoid the previous bottleneck.
```

### Escaped and unescaped name in single buffer

Another variant of the above.
The escaped and unescaped name are stored in a single buffer.
This avoids extra allocations and extra pointers and sizes in structs.

The escaped name could also be stored lazily only when needed.
This would solve the cleanup problem.

While this may seem like the ideal solution, there are still some downsides.
The biggest problem is the API design.
If the API uses escaped names a lot (because it is more convenient), then this essentially degrades into the "Both escaped and unescaped name" solution.
Even if APIs exist for both escaped and unescaped names, the convenience benefit, will lead to more use of escaped names.
This means the escaped name will be stored for many keys and therefore the benefit of the lazy allocation is negated.

Without the lazy allocation benefit, the only difference to "Both escaped and unescaped name" is that we have fewer pointers and sizes in structs.
This saves some amount of memory and allocations, but makes internal code more difficult to write and understand.

## Decision

Go with "Only unescaped name, with separate namespace" from above:

> - Store only unescaped name inside `struct _Key`

Reviewer: ```suggestion
- Store only unescaped name with size inside `struct _Key`
```
